File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        std::unordered_map<Node*, Node*> cur_copy;
        Node* cur = head;
        while (cur) {
            Node* node = new Node(cur->val);
            cur_copy[cur] = node;
            cur = cur->next;
        }
        cur = head;
        while (cur) {
            Node* node = cur_copy[cur];
            node->next = cur_copy[cur->next];
            node->random = cur_copy[cur->random];
            cur = cur->next;
        }
        return cur_copy[head];
    }
};

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        node_to_clone = {None: None}
        curr = head
        while curr:
            clone = Node(curr.val)
            node_to_clone[curr] = clone
            curr = curr.next
        curr = head
        while curr:
            clone = node_to_clone[curr]
            clone.next = node_to_clone[curr.next]
            clone.random = node_to_clone[curr.random]
            curr = curr.next
        return node_to_clone[head]